src/nvim/generators/gen_api_ui_events.lua

```diff
@@ -50,6 +51,106 @@ function write_arglist(output, ev, need_copy)
   end
 end
 
+function extract_and_write_arglist(output, ev)
+  local hlattrs_args_count = 0
+  for j = 1, #ev.parameters do
+    local param = ev.parameters[j]
+    local kind = param[1]
+    output:write('  '..kind..' arg_'..j..' = ')
+    if kind == 'HlAttrs' then
+      -- The first HlAttrs argument is rgb_attrs and second is cterm_attrs
+      output:write('dict2hlattrs(args.items['..(j-1)..'].data.dictionary, '..(hlattrs_args_count == 0 and 'true' or 'false')..');\n')
+      hlattrs_args_count = hlattrs_args_count + 1
+    elseif kind == 'Object' then
+      output:write('args.items['..(j-1)..'];\n')
+    else
+      output:write('args.items['..(j-1)..'].data.'..string.lower(kind)..';\n')
+    end
+  end
+end
+
+function call_ui_event_method(output, ev)
+  output:write('  ui_call_'..ev.name..'(')
+  for j = 1, #ev.parameters do
+    output:write('arg_'..j)
+    if j ~= #ev.parameters then
+      output:write(', ')
+    end
+  end
+  output:write(');\n')
+end
+
+function make_raw_line_and_call(output)
+  -- arg_1 = grid, arg_2 = row, arg_3 = startcol, arg_4 = cells, arg_5 = endcol, arg_6 = clearcol, arg_7 = clearattr
+  
+  output:write([[
+  Integer arg_5, arg_6, arg_7;
+  LineFlags lineflags = kLineFlagWrap;
+  schar_T *chunk;
+  sattr_T *attrs;
+  size_t size_of_cells = arg_4.size;
+  size_t no_of_cells = size_of_cells;
+  arg_5 = arg_3;
+
+  // checking if clearcol > endcol
+  if (!STRCMP(arg_4.items[size_of_cells-1].data.array.items[0].data.string.data, " ")) {
+    no_of_cells = size_of_cells - 1;
+  }
+
+  // getting endcol
+  for (size_t i = 0; i < no_of_cells; i++) {
+    arg_5++;
+    if (arg_4.items[i].data.array.size == 3) {
+      arg_5 += arg_4.items[i].data.array.items[2].data.integer - 1;
+    }
+  }
+
+  size_t ncells = (size_t)(arg_5 - arg_3);
+  chunk = xmalloc(ncells * sizeof(schar_T) + 1);
+  attrs = xmalloc(ncells * sizeof(sattr_T) + 1);
+
+  size_t j = 0;
+  size_t k = 0;
+  for (size_t i = 0; i < no_of_cells; i++) {
+    STRCPY(chunk[j++], arg_4.items[i].data.array.items[0].data.string.data);
+    if (arg_4.items[i].data.array.size == 3) {
+      // repeat present
+      for (size_t i_intr = 1; i_intr < (size_t)arg_4.items[i].data.array.items[2].data.integer; i_intr++) {
+        STRCPY(chunk[j++], arg_4.items[i].data.array.items[0].data.string.data);
+        attrs[k++] = (sattr_T)arg_4.items[i].data.array.items[1].data.integer;
+      }
+    } else if (arg_4.items[i].data.array.size == 2) {
+      // repeat = 1 but attrs != last_hl
+      attrs[k++] = (sattr_T)arg_4.items[i].data.array.items[1].data.integer;
+    }
+    if (j > k) {
+      // attrs == last_hl
+      attrs[k] = attrs[k-1];
+      k++;
+    }
+  }
+
+  if (!STRCMP(arg_4.items[size_of_cells-1].data.array.items[0].data.string.data, " ")) { 
+    arg_7 = arg_4.items[size_of_cells-1].data.array.items[1].data.integer;
+    arg_6 = arg_5 + arg_4.items[size_of_cells-1].data.array.items[2].data.integer;
```

Comment:
I think we are overcounting the cleared area here. `arg_5` is already the maximum width before this, so we would set `arg_6` to it and instead reduce `arg_5` by the cleared amount.